### PR TITLE
add bug fix for multiple nested functions as params

### PR DESCRIPTION
**Pull request by bohandley**
**What is this feature?**
A bug fix for nested params that are seriesByTag functions in the asPercent function.
Fixes: https://github.com/grafana/support-escalations/issues/5600

When using the code editor and entering this query, a user may switch to the query builder. This switch reads the second param as a function and pulls out of the asPercent function into its own function. It is now incorrect. If a user chooses new tags or changes the query and then switches back to the code editor, this parsing issue is exposed and it parses the incorrectly parsed `seriesByTag` into the incorrect place in the code query. See escalation.

Here is an example of the `asPercent` function with two params that are `seriesByTag` functions and how to reproduce the bug.
```
asPercent( seriesByTag('name=linux_sysctl_fs.file-nr', 'host=staging-zh-swarm-fs-210'), seriesByTag('name=linux_sysctl_fs.file-max', 'host=staging-zh-swarm-fs-210') )
```
1. enter this into the code editor as a string
2. switch to the query builder by clicking the pencil
3. change a tag value in the series component
4. switch back to the code editor
5. without the fix the query is now incorrectly formatted and the second param function is nested in the wrong place like the following.
```
asPercent(seriesByTag(seriesByTag('name=linux_sysctl_fs.file-nr', 'host=staging-zh-swarm-fs-210'), 'name=linux_sysctl_fs.file-max', 'host=staging-zh-swarm-fs-210') )
```

**Why do we need this feature?**
While a user can enter two function type params in the asPercent function in code, the query parser was not meant to handle two functions as params. This is an edge case and would require extensive refactoring of the parser so we offer the following bug fix. 

This fix keep the second query as a string and keeps it inside the `asPercent` function so that a user can update a tag or update the query and switch back to the code editor and have no parsing issues.

**Special notes for your reviewer:**
Follow instructions above.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/plugins/datasource/graphite/graphite_query.ts [hunk 1]
```diff
@@ -326,3 +330,44 @@ export default class GraphiteQuery {
...
+        return {
+          type: 'string',
+          value: `${p.name}(${stringParams})`,
+        };
+      }
+
+      return p;
+    });
+  }
+}
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/graphite/graphite_query.ts:373]: Here's some types to avoid eslint-ignores
```suggestion

type ASTNodeParams = {
  type: string
  name?: string
  value: string
  params?: ASTNodeParams[]
}

interface ASTNode {
  params: ASTNodeParams[]
}

/**
 * mutates the second seriesByTag function into a string to fix a parsing bug
 * @param astNode
 * @param innerFunc
 */
function handlemultipleSeriesByTagsParams(astNode: ASTNode, innerFunc: any) {
  // if function is asPercent and has two params that are function seriesByTags keep the second as a string otherwise we have a parsing error
  if (innerFunc.def.name === 'asPercent' && astNode.params.length >= 2) {
    let count = 0;
    astNode.params = astNode.params.map((p: ASTNodeParams) => {
      if (p.type === 'function') {
        count += 1;
      }

      if (count === 2 && p.type === 'function' && p.name === 'seriesByTag') {

        // convert second function to a string
        const stringParams = p.params?.reduce(
          (acc: string, p: ASTNodeParams, idx: number, paramsArr: any[]) => {
            if (idx === 0 || idx !== paramsArr.length - 1) {
              return `${acc}'${p.value}',`;
            }

            return `${acc}'${p.value}'`;
          },
          ''
        );

        return {
          type: 'string',
          value: `${p.name}(${stringParams})`,
        };
      }

      return p;
    });
  }
}
```

Comment (gtk-grafana) [reply public/app/plugins/datasource/graphite/graphite_query.ts:373]: kinda rough on formatting, and the types should go to the top of the file, but lemme know what you think

Comment (gtk-grafana) [reply public/app/plugins/datasource/graphite/graphite_query.ts:373]: even if these definitions aren't entirely complete, if we use these more and more over time, we'll update the definitions and they should get closer to being correct

Comment (bohandley) [reply public/app/plugins/datasource/graphite/graphite_query.ts:373]: Cool! @gtk-grafana I have this PR up for the types https://github.com/grafana/grafana/pull/67067 which adds a little more to the ast node definition. Can we use this? We'll need even more to make it applicable to the other areas where ast nodes are applied to parts of the query, but it's a work in progress, a little at a time.

Comment (gtk-grafana) [reply public/app/plugins/datasource/graphite/graphite_query.ts:373]: Yes please, let's merge that on in! I'd like to avoid adding any `any` or eslint ignores, even if we need to do partial refactors like this or just define types for new code as I suggested.

Comment (bohandley) [reply public/app/plugins/datasource/graphite/graphite_query.ts:373]: I think this is a good goal! to avoid using `any` or eslint ignores is the optimum. I also want to mention, there is a price to refactoring legacy code that is littered with `any`. This took me a day to refactor and broke almost every test in the parser. I agree we should try to implement a solution like yours but also be aware that it can require a significant time investment. When it comes to investing time, I think we should bring issues like these up in quarterly planning.

File: public/app/plugins/datasource/graphite/graphite_query.ts [hunk 2]
```diff
@@ -326,3 +330,44 @@ export default class GraphiteQuery {
...
+
+/**
+ * mutates the second seriesByTag function into a string to fix a parsing bug
+ * @param astNode
+ * @param innerFunc
+ */
+// eslint-disable-next-line
+function handlemultipleSeriesByTagsParams(astNode: any, innerFunc: any) {
+  // if function is asPercent and has two params that are function seriesByTags keep the second as a string otherwise we have a parsing error
+  if (innerFunc.def.name === 'asPercent' && astNode.params.length >= 2) {
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/graphite/graphite_query.ts:342]: so I was able to repro this issue with other functions besides asPercent, like consolidateBy. I know very little about graphite functions, so if you can take a look and make sure there aren't any other functions that take multiple param that could get formatted weird like this that would make me feel better.

Tl;DR https://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.aliasByTags, could this impact all graphite functions that take a `seriesList`?

**Comment by gtk-grafana** [comment]
I can verify this fixes the provided example, but not sure if there are other functions besides asPercent that could get formatted all wonky like this.

File: public/app/plugins/datasource/graphite/graphite_query.ts (hunk 2 above)
Comment (bohandley) [reply public/app/plugins/datasource/graphite/graphite_query.ts]: Scratch that, I'm going to have to check all 105 functions for this 😦

**Comment by bohandley** [comment]
> I can verify this fixes the provided example, but not sure if there are other functions besides asPercent that could get formatted all wonky like this.

So far, the first 20 functions I’ve tried break when switching from code to builder, updating a tag in builder, switching back to code, parsing two seriesAsTags functions as parameters. I will continue to test the other 85 next week when I am back from vacation.

Not all of the functions are meant to have multiple parameters though, but there is no check on either builder or code for how many you put in. The main check is querying the function and getting the error back from Graphite, whether the function accepts multiple arguments. 

One option is to handle all functions that contain two parameters that are seriesAsTags functions with this fix. That could cover everything relating to editing the tag when switching back and forth between code and builder. After checking the rest of the functions, I am inclined to have this be the solution and add tests to cover this, as you suggested, @gtk-grafana WDYT?

**Comment by bohandley** [comment]
All functions are affected by this bug, so I am adapting the solution to be not specific for the asPercent function. I have added a test that checks that seriesByTags functions params are not wrapped, which was the original bug. This test calls `updateModelTarget` which is where the bug is expressed. All functions [are reduced here](https://github.com/grafana/grafana/blob/4ee894b0bea4306bc171a16c0cc5f15fd799b414/public/app/plugins/datasource/graphite/graphite_query.ts#L179-L188) which is a good case for refactoring. All functions in the query model are wrapped inside out, and so multiple params that are functions cannot be treated this way.

**Review by gtk-grafana — APPROVED** [review]
Awesome work on this! Thanks for taking the time to find a more robust solution!!
